pss/trojan_chunk.go

```diff
@@ -28,173 +28,180 @@ import (
 	}
 
 	// hash trojan chunk fields with different nonces until an acceptable one is found
 	// TODO: prevent infinite loop
 	for {
-		s, _ := serializeTrojanChunk(span, nonce, m) // err always nil here
-		hash, hashErr := hashTrojanChunk(s)
-		if hashErr != nil {
-			return emptyAddr, emptyPayload, hashErr
+		s := append(append(span, nonce...), m...) // serialize trojan chunk fields
+		hash, err := hash(s)
+		if err != nil {
+			return nil, err
 		}
 
 		// take as much of the hash as the targets are long
-		if hashPrefixInTargets(hash[:targetsLength], targets) {
-			// if nonce found, stop loop and return matching hash as address
-			return hash, s, nil
+		if contains(targets, hash[:targetsLen]) {
+			// if nonce found, stop loop and return chunk
+			return chunk.NewChunk(hash, s), nil
 		}
 		// else, add 1 to nonce and try again
-		// TODO: test loop-around
 		nonceInt.Add(nonceInt, big.NewInt(1))
-		nonce = nonceInt.Bytes()
+		// loop around in case of overflow
+		if nonceInt.BitLen() > 256 {
+			nonceInt = big.NewInt(0)
+		}
+		nonce = padBytes(nonceInt.Bytes()) // pad in case Bytes call is not long enough
 	}
 }
 
-// serializeTrojanChunk appends the span, nonce and trojan message fields of a trojan chunk and returns the result
-// this can be used as the input for trojan chunk hash calculation
-func serializeTrojanChunk(span, nonce, message []byte) ([]byte, error) {
-	h := append(span, nonce...)
-	s := append(h, message...)
-	return s, nil
-}
-
-// hashTrojanChunk hashes the serialization of trojan chunk fields with the trojan hashing func
-func hashTrojanChunk(s []byte) ([]byte, error) {
-	trojanHashingFunc.Reset() // TODO: why do we need to do this?
-	if _, err := trojanHashingFunc.Write(s); err != nil {
-		return []byte{}, err
+// hash hashes the serialization of trojan chunk fields with the trojan hashing func
+func hash(s []byte) ([]byte, error) {
+	hasher := hashFunc()
+	hasher.Reset()
+	if _, err := hasher.Write(s); err != nil {
+		return nil, err
 	}
-	return trojanHashingFunc.Sum(nil), nil
+	return hasher.Sum(nil), nil
 }
 
-// hashPrefixInTargets returns whether the given hash prefix appears in the targets given as a collection of byte slices
-func hashPrefixInTargets(hashPrefix []byte, targets [][]byte) bool {
-	for i := range targets {
-		if bytes.Equal(hashPrefix, targets[i]) {
+// contains returns whether the given collection contains the given elem
+func contains(col [][]byte, elem []byte) bool {
+	for i := range col {
+		if bytes.Equal(elem, col[i]) {
 			return true
 		}
 	}
 	return false
 }
 
-// MarshalBinary serializes a trojanMessage struct
-func (tm *trojanMessage) MarshalBinary() (data []byte, err error) {
+// padBytes adds 0s to the given byte slice as left padding,
+// returning this as a new byte slice with a length of exactly 32
+// given param is assumed to be at most 32 bytes long
+func padBytes(b []byte) []byte {
+	l := len(b)
+	if l == 32 {
+		return b
+	}
+	bb := make([]byte, 32)
+	copy(bb[32-l:], b)
```

Comment:
i don't think endianness matters for the nonce, since we are only interpreting it as a number to add 1 to it. 

keep in mind `big.Int`'s `Bytes` func interprets slices as big-endian number representations (as does `SetBytes`). this, in turn, causes the last byte in the array (number 32) to have its value increased when we add 1 to the nonce as an integer.

i anyways added a `TestPadBytes` function. please check it out and let me know if you see anything that doesn't add up.